pkg/playback/http_playback_test.go

```diff
@@ -0,0 +1,275 @@
+	server := httpRecorder.initializeServer(addressString)
+	go func() {
+		if err := server.ListenAndServe(); err != http.ErrServerClosed {
+			// unexpected error
+			log.Fatalf("ListenAndServe(): %v", err)
+		}
+	}()
+
+	// Send it 2 requests
+	res1, err := http.Get("http://localhost:8080/")
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse1.StatusCode, res1.StatusCode)
+
+	res2, err := http.Get("http://localhost:8080/")
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse2.StatusCode, res2.StatusCode)
+
+	// Shutdown record server
+	resShutdown, err := http.Get("http://localhost:8080/playback/stop")
+	server.Shutdown(context.TODO())
+	assert.NoError(t, err)
+	defer resShutdown.Body.Close()
+
+	// --- Set up a replay server
+	httpReplayer := newReplayServer(webhookURL)
+	err = httpReplayer.readCassette(&cassetteBuffer)
+	check(t, err)
+
+	replayServer := httpReplayer.initializeServer(addressString)
+	go func() {
+		if err := replayServer.ListenAndServe(); err != http.ErrServerClosed {
+			// unexpected error
+			log.Fatalf("ListenAndServe(): %v", err)
+		}
+	}()
+
+	// Send it the same 2 requests:
+	replay1, err := http.Get("http://localhost:8080/")
+	assert.NoError(t, err)
+	replay2, err := http.Get("http://localhost:8080/")
+	assert.NoError(t, err)
+
+	// Assert that the original responses and replayed responses are the same
+	check(t, assertHTTPResponsesAreEqual(t, res1, replay1))
+	check(t, assertHTTPResponsesAreEqual(t, res2, replay2))
+
+	// Also sanity check that the mock server is responding with the expected responses
+	assert.Equal(t, "testHeaderValue", res1.Header.Get("testHeader"))
+	bodyBytes1, err := ioutil.ReadAll(res1.Body)
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse1.Body, bodyBytes1)
+
+	bodyBytes2, err := ioutil.ReadAll(res2.Body)
+	assert.NoError(t, err)
+	assert.Equal(t, mockResponse2.Body, bodyBytes2)
+
+	// Shutdown replay server
+	replayServer.Shutdown(context.TODO())
+}
+
+// Test the full server by switching between modes, loading and ejecting cassettes, and sending real stripe requests
+func TestPlaybackSingleRunCreateCustomerAndStandaloneCharge(t *testing.T) {
+	var remoteURL string
+
+	// Initialize the mock responses we want to serve
+	mockResponse1 := httpResponse{}
+	mockResponse1.StatusCode = 200
+	mockResponse1.Headers = make(http.Header)
+	mockResponse1.Headers.Add("testHeader", "testHeaderValue")
+	mockResponse1.Body = []byte("body 1")
+
+	mockResponse2 := httpResponse{}
+	mockResponse2.StatusCode = 402
+	mockResponse2.Headers = make(http.Header)
+	mockResponse2.Body = []byte("body 2")
+	fixtureResponses := []httpResponse{mockResponse1, mockResponse2}
+
+	ts := startMockServer(fixtureResponses)
+	defer ts.Close()
+	remoteURL = ts.URL
+
+	// -- Setup Playback server
+	addressString := "localhost:13111"
+	cassetteFilepath := "test_record_replay_single_run.yaml"
+
+	// for now, write cassettes to this directory. Ideally we have a test output folder
+	cassetteDirectory, err := filepath.Abs("")
```

Comment:
We shouldn't do this.
This would be better: https://golang.org/pkg/io/ioutil/#TempDir

Or build in the ability to inject a mock filesystem (i.e. define a `FileOpener` interface and put it on the server struct and delegate to that) -- but I think tempdir gives higher confidence since this is an integration test more than a unit test.

Reply:
Fixed in d253aa567c91a011f30b1a8024ed988a4dae034e.